daemon/signals.c

```diff
@@ -157,6 +229,11 @@ void signals_handle(void) {
                             case NETDATA_SIGNAL_FATAL:
                                 fatal("SIGNAL: Received %s. netdata now exits.", name);
 
+                            case NETDATA_SIGNAL_CHILD:
+                                if (reaper_enabled == 1)
+                                    reap_child();
```

Comment:
I think that at this point, there may have been multiple `SIGCHLD` signals before `signals_waiting[i].count = 0;` was set.

I think a single `reap_child()` call is not enough and that we need a loop. What do you think?

Reply:
Yes indeed, didn't spot this wasn't the actual signal handler.

I would prefer to actual rework the signal handlers to be inline, otherwise you've going to have to deal with races. This would should also be more performant thoughts?

Reply:
I'm afraid that is not going to work because in the signal handlers themselves you can do almost nothing, and netdata already does **a lot** of heavy lifting in `signals_handle()`. You cannot even use mutexes such as those from `myp_reap()` in `reap_child()`. See [here](http://man7.org/linux/man-pages/man7/signal-safety.7.html) for reference.

`signals_handle()` is being called by the netdata daemon master thread, that is possibly the one with `pid == 1` in the container case, and it's devoted to exclusively handle signals after the initialization. Due to the limitations of signal safety the scope of a perfect solution would be too big in relation to the motivation behind this pull request.

I would prefer you figure out  a way to catch the signals without making a lot of changes.

Reply:
Hmm actually looks like this was a conscious decision, to avoid long running signal handlers. I will take a different approach.

Reply:
Yep already spotted that when investigating the options, but thanks for the heads up.